### ISSUE TITLE
[HOLD #47534][$250]  Xero - Strange animation after dismissing 2FA setup RHP and not here page on LHN

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#45572]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.8-1
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**Found when validating PR :** https://github.com/Expensify/App/pull/44059
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
Precondition:
- User has a workspace.
- The workspace is connected to Xero.
- User has not set up 2FA.

1. Go to staging.new.expensify.com
2. Log out if logged in.
3. Log in to the account from precondition.
4. Click Enable two-factor authentication.
5. Click the back button on 2FA set up RHP.
6. Navigate between Account settings, Search and Inbox.

## Expected Result:
App should throw 2FA set up modal again if user exits 2FA set up flow.

## Actual Result:
In Step 5, when clicking on back button on 2FA setup RHP, the RHP dismissal has a weird animation.
In Step 6, after switching between sections, not here page shows up on the left hand panel.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/f5c50571-fcb7-453e-b0b5-dd15cf31ec23

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e2d69ad81006978f</li>
        <li>Upwork Job ID: 1813578824903964216</li>
        <li>Last Price Increase: 2024-09-04</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @ahmedGaber93</details>

**Comment by izarutskaya** [comment]
We think this issue might be related to the #collect project.

**Comment by MonilBhavsar** [comment]
Commented on the offending PR https://github.com/Expensify/App/pull/44059

**Comment by rushatgabhane** [comment]
The issue is present after reverting this PR too. its unrelated to the PR and related to navigation

**Comment by MonilBhavsar** [comment]
Ok thanks for checking!

**Comment by MonilBhavsar** [comment]
> The issue is present after reverting this PR too. 

@rushatgabhane if we revert this PR, then the 2FA modal doesn't show up and looks like issue is not reproducible?

**Comment by rushatgabhane** [comment]
@MonilBhavsar let me rephrase what i mean - this same issue is present on production also. The PR above did not make any changes to how back navigation was handled

1. Go to https://new.expensify.com/settings/security/two-factor-auth?backTo=home
2. Click back

https://github.com/user-attachments/assets/a228ff3e-5722-40cb-adb2-a0c31277ac4d

**Comment by MonilBhavsar** [comment]
Ok thanks for clarifying. Makes sense

**Comment by rushatgabhane** [comment]
im happy to help, but i don't have the fix yet

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Dismissing 2FA RHP gives weird animation and shows not here page on LHN.

### What is the root cause of that problem?
When we press back, we will use the `backTo` as the fallback route.
https://github.com/Expensify/App/blob/c82183bca6338758888ca9e43591fc1f11234928/src/pages/settings/Security/TwoFactorAuth/Steps/CodesStep.tsx#L57

The fallback route is set to HOME.
https://github.com/Expensify/App/blob/c82183bca6338758888ca9e43591fc1f11234928/src/Expensify.tsx#L272

Using HOME as the fallback route will lead to the weird animation and it's a known issue which is previously fixed in https://github.com/Expensify/App/pull/36050 where we remove all HOME from the fallback route.

### What changes do you think we should make in order to solve the problem?

Since users can manually edit the `backTo` to home as shown in the comment [here](https://github.com/Expensify/App/issues/45572#issuecomment-2233438095), I think the best solution is to set the fallback to an empty string if it's HOME.
```
if (fallbackRoute === ROUTES.HOME) {
    fallbackRoute = '';
}
```

_or if the fallback route is HOME, we [pop every screen on the stack](https://github.com/Expensify/App/blob/c82183bca6338758888ca9e43591fc1f11234928/src/libs/Navigation/Navigation.ts#L205)_

**Comment by ahmedGaber93** [comment]
@bernhardoj Thanks for the proposal.

Both your solutions not back me to home page. The first back me to `settings/security`, and the second using `StackActions.popToTop()` back to `settings/profile`. Did I missed something?

**Comment by rushatgabhane** [comment]
In this case, we want to take the user to home page

**Comment by ahmedGaber93** [comment]
> Do we want to navigate the user back from the 2FA page?

Yes, I think that.

> are we fixing the animation issue somewhere?

The issue in not only in animation, the center pane also seems empty because the route `/home` doesn't redirect to last opened report

**Comment by bernhardoj** [comment]
> the center pane also seems empty because the route /home doesn't redirect to last opened report

Our navigation currently prioritizes the UP type which will replace the current route with the destination route. In this case, the 2FA RHP is replaced with HOME, which is a BOTTOM TAB SCREEN (LHN), that's why the report screen is blank.

```
if (type === 'UP') // change the action to REPLACE
else if (destination === bottom tab screen) {
    // 1. Push the bottom tab screen
    // 2. Push the matched central pane screen, in this case, a report screen
}
```

Because we want the HOME to show when the backTo is home, we can fix this by moving down the `type === 'UP' condition and also handle it inside the bottom tab condition.

```
if (destination === bottom tab screen) {
    // 1. Push the bottom tab screen
    // 2. If type is UP, replace the current screen with the matched central pane screen
    // Else, push the matched central pane screen
} else if (type === 'UP) // change the action to REPLACE
```

However, there is another issue. After navigating to the LHN/HOME, pressing the settings page again will show a blank profile page. The profile page exists and is active in the nav stack, but it is just blank and you can see the tooltip never hides and there is no hover color at all. I remember seeing the blank issue that is caused by the react-compiler, but I can't find the issue link.

https://github.com/user-attachments/assets/0cf4fc9c-c95b-4d63-880f-9f9c298f40fb

**Comment by rushatgabhane** [comment]
i suggest that we hold on

**Comment by ahmedGaber93** [comment]
> i can't reproduce on it

https://github.com/Expensify/App/pull/46087 was merged into main, and the issue is still reproduced on latest main

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by ahmedGaber93** [comment]
Waiting on proposals

**Comment by trjExpensify** [comment]
Are we not evaluating @bernhardoj's?

**Comment by ahmedGaber93** [comment]
Sorry, I missed giving feedback on the last comment.

**Comment by ahmedGaber93** [comment]
>  I remember seeing the blank issue that is caused by the react-compiler, but I can't find the issue link.

@bernhardoj Do you found the issue link? We need to confirm the blank profile screen is fixed and not a regression from your solution.

**Comment by bernhardoj** [comment]
I can't find the one that I meant, but there is a new issue created that is very similar https://github.com/Expensify/App/issues/47041

**Comment by ahmedGaber93** [comment]
@bernhardoj I retest your fix again today, and it seems the blank issue is not reproduced now with your fix and last main. Can you confirm?
The tooltip freezing issue still reproduced, but it also reproduced on the last main if you keep "security" screen opened and jump to tab "inbox" then tab "profile settings"

Your fix now works fine, can you please test it well, and it doesn't cause any errors on other navigations, then update your proposal?

**Comment by ahmedGaber93** [comment]
@bernhardoj bump for https://github.com/Expensify/App/issues/45572#issuecomment-2278059375

**Comment by bernhardoj** [comment]
> and it seems the blank issue is not reproduced now with your fix and last main. Can you confirm?

Yeah, same.

> Your fix now works fine, can you please test it well, and it doesn't cause any errors on other navigations, then update your proposal?

Since the changes affect navigating to BOTTOM TAB screens with the type of UP, I searched for all usages of `goBack` with the fallback route of each bottom tab route (SETTINGS, HOME, SEARCH) or `navigate` with a type of UP.

For HOME, there is another usage other than what we have in this issue.
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/libs/Navigation/AppNavigator/Navigators/OnboardingModalNavigator.tsx#L47-L49

But the fallback route is never used because we pass `shouldPopToTop` as true.
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/libs/Navigation/Navigation.ts#L198-L204

For SETTINGS, 

1. About page
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/pages/settings/AboutPage/AboutPage.tsx#L138
The fallback route is never used because `isFirstRouteInNavigator` is false.
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/libs/Navigation/Navigation.ts#L222-L225

2. Troubleshoot page
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/pages/settings/Troubleshoot/TroubleshootPage.tsx#L113
Same as above

3. Wallet empty state
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/pages/settings/Wallet/WalletEmptyState.tsx#L44
Same as above

For SEARCH,

2 usages in the search page itself, both for large (SearchPage) and small screen (SearchPageBottomTab)
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/pages/Search/SearchPage.tsx#L38-L41
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/pages/Search/SearchPageBottomTab.tsx#L54-L56

But I don't know how to trigger the not found page here.

However, when I test to open the 2FA and change the `backTo` to `settings`, it does nothing. It's because, if we are trying to navigate to the same bottom tab screen, it will do nothing.
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/libs/Navigation/linkTo/index.ts#L153-L157
https://github.com/Expensify/App/blob/86da342bdc00960903cb5a5b75730f8b07b7f667/src/libs/Navigation/linkTo/getActionForBottomTabNavigator.ts#L36-L38

Maybe we can ask SWM to add support of fallback route to bottom tab screen.

**Comment by ahmedGaber93** [comment]
@bernhardoj Thanks for all the context and details.

> Maybe we can ask SWM to add support of fallback route to bottom tab screen.

I think we need to do that. Although using `push` instead `replace` will display the correct route, it will push a new route, and if the user click on browser arrow back button, it will back again to 2FA modal

 

https://github.com/user-attachments/assets/6d1d6422-d2de-46bd-97b1-daeaf0e7a03c

@MonilBhavsar what do you think of asking SWM to handle this?

**Comment by ahmedGaber93** [comment]
@VictoriaExpensify @MonilBhavsar we don't have accepted proposals yet, I think we can share this issue with SWM instead of waiting for new proposals, They always handle the navigation issues.

**Comment by ahmedGaber93** [comment]
Waiting on proposals.

**Comment by ahmedGaber93** [comment]
> @VictoriaExpensify @MonilBhavsar we don't have accepted proposals yet, I think we can share this issue with SWM instead of waiting for new proposals, They always handle the navigation issues.

CC @adamgrzybowski

**Comment by MonilBhavsar** [comment]
> what do you think of asking SWM to handle this?

Let's get some thoughts 👍  I see you have tagged Adam Grzybowski

**Comment by adamgrzybowski** [comment]
Hey guys, we are currently working on some changes to the navigation that will influence how the `goBack` works. My question is how urgent is this fix?  We can work on some solution, but it will probably be discarded after these changes. 

BTW Maybe as a simple fix, we can just not use the `HOME` as a param for the `goBack` function? We could use a combination of `dismissModal` and `navigate` to `HOME`?

**Comment by ahmedGaber93** [comment]
Not overdue, we are still discussing the solution

**Comment by ahmedGaber93** [comment]
@MonilBhavsar What do you think https://github.com/Expensify/App/issues/45572#issuecomment-2307184808

**Comment by VictoriaExpensify** [comment]
Not overdue

**Comment by ahmedGaber93** [comment]
@MonilBhavsar What do you think https://github.com/Expensify/App/issues/45572#issuecomment-2307184808?

**Comment by VictoriaExpensify** [comment]
bump @adamgrzybowski :

> Yeah I don't think this is urgent @adamgrzybowski . I think it makes sense to put this on the back-burner and revisit it once the other navigation changes are in place. Do you have a GH for those changes that we can reference here?

**Comment by trjExpensify** [comment]
>we are currently working on some changes to the navigation that will influence how the goBack works. My question is how urgent is this fix? We can work on some solution, but it will probably be discarded after these changes.

What changes are these, and can you link an issue? I don't think this is super pressing, but If we're putting it on hold, let's make sure that's reflected in the issue title.

**Comment by VictoriaExpensify** [comment]
Ok I think this is the issue @adamgrzybowski was referencing - https://github.com/Expensify/App/issues/47534

He's gone ooo and that issue has been changed to monthly, so I think we're safe to put this on hold for now

**Comment by VictoriaExpensify** [comment]
Not OD, this is on hold

**Comment by VictoriaExpensify** [comment]
As above. Making this a monthly as that's what the related issue is

**Comment by NJ-2020** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Xero - Strange animation after dismissing 2FA setup RHP and not here page on LHN
### What is the root cause of that problem?
First let split this problem into 2 issues.
- Why the sidebar had strange animation moving from right to left 
- Why only sidebar is displaying and where is the main chat?

1. When we have a backTo param inside our url which is `HOME`, currently we're showing Sidebar component for HOME page inside BottomTabNavigator.tsx file
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/AppNavigator/Navigators/BottomTabNavigator.tsx#L26-L29
But why we see a strange animation for Home sidebar since we disabled it here: https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/AppNavigator/Navigators/BottomTabNavigator.tsx#L16-L19
Because we invoked the BottomTabNavigator inside AuthScreens and enabled the animation here
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/AppNavigator/AuthScreens.tsx#L223
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/AppNavigator/AuthScreens.tsx#L412-L416
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/AppNavigator/getRootNavigatorScreenOptions.ts#L125-L127
We can see here that the animation is enabled, I've tried disabling the animation and it solved the first issue (sidebar strange animation)
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/AppNavigator/getRootNavigatorScreenOptions.ts#L21-L27

2. Why the main chat is not displaying?, because we use `getMatchingCentralPaneRouteForState` function to display the central pane route for bottom tab navigator i.e for HOME page we will display Report chat.

And for navigation i think we use `linkTo` function and we invoke the  `getMatchingCentralPaneRouteForState` inside the `linkTo` function if it does not match these 3 conditions
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/linkTo/index.ts#L95
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/linkTo/index.ts#L122
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/linkTo/index.ts#L130
If the above 3 conditions are false we will invoke the `getMatchingCentralPaneRouteForState` inside this condition
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/linkTo/index.ts#L151

But when pressing the go back button we use the `goBack` function which uses the `CONST.NAVIGATION.TYPE.UP` type which leads this condition to true
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/linkTo/index.ts#L122
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/Navigation.ts#L235 and [here](https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/Navigation/Navigation.ts#L245)
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/actions/TwoFactorAuthActions.ts#L21-L24
and will not invoke the `getMatchingCentralPaneRouteForState` function
### What changes do you think we should make in order to solve the problem?
Instead of using `goBack` function we can use `navigate` here
`Navigation.navigate(backTo)`
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/libs/actions/TwoFactorAuthActions.ts#L21-L24

### Result

https://github.com/user-attachments/assets/3fce7d44-c741-4358-966a-8f186c3625be

### What alternative solutions did you explore? (Optional)

**Comment by NJ-2020** [comment]
@ahmedGaber93 Can you please kindly take a look at my [proposal](https://github.com/Expensify/App/issues/45572#issuecomment-2381383755)

**Comment by ahmedGaber93** [comment]
@NJ-2020 use `navigate` instead of `goBack` is already proposed by adamgrzybowski here https://github.com/Expensify/App/issues/45572#issuecomment-2307184808 as a workaround, but we decide to wait for fixing `goBack` https://github.com/Expensify/App/issues/45572#issuecomment-2316788510.

**Comment by ahmedGaber93** [comment]
@VictoriaExpensify I think we need to add monthly label because we hold it for monthly issue https://github.com/Expensify/App/issues/47534

**Comment by ahmedGaber93** [comment]
@VictoriaExpensify bump https://github.com/Expensify/App/issues/45572#issuecomment-2397601319

**Comment by MonilBhavsar** [comment]
Adjusted label, thanks!

**Comment by trjExpensify** [comment]
I don't think this should be closed, it's held on https://github.com/Expensify/App/issues/47534 so I've updated the title to reflect that better to follow the breadcrumbs.

**Comment by ahmedGaber93** [comment]
https://github.com/Expensify/App/pull/49539 was deployed to production 4 days ago https://github.com/Expensify/App/pull/49539#issuecomment-2662467810, the navigation issue looks fixed

https://github.com/user-attachments/assets/41bc8437-9f0e-40ba-92b1-4c3205c50d0a

After the navigation issue was fixed, we can focus on the expected Result now
> App should throw 2FA set up modal again if user exits 2FA set up flow.

@VictoriaExpensify Can you please confirm the expected Result?

**Comment by ahmedGaber93** [comment]
@VictoriaExpensify bump https://github.com/Expensify/App/issues/45572#issuecomment-2673136703

**Comment by ahmedGaber93** [comment]
@VictoriaExpensify is OOO until March 16th.

@trjExpensify could you please confirm the expected Result? 
>  App should throw 2FA set up modal again if user exits 2FA set up flow.

In this case, user must complete the 2FA set up to be able to use the App

**Comment by trjExpensify** [comment]
@ahmedGaber93 we're making this flow a lot better in [this issue](https://github.com/Expensify/App/issues/57684), so I think we can do nothing here.

**Comment by ahmedGaber93** [comment]
Great! we should close this then.

**Comment by trjExpensify** [comment]
Done!
